### PR TITLE
feat: initial keycloak integration

### PR DESCRIPTION
**Pull request by yassinedorbozgithub**
This PR is a paring implementation done by both @IkbelTalebHssan and @yassinedorbozgithub 

It includes :
- Integrating Keycloak and Postgres services
- Using next-auth package
- Adding login / logout buttons, name of connected user
- Automating Keycloak configs
- Enabling and customization self registration form

## Summary by CodeRabbit

- **New Features**
	- Introduced a Keycloak authentication integration for improved user management and security.
	- Added user login and logout components for a seamless authentication experience.
	- Enhanced the header to conditionally display login/logout options based on user session status.
	- Implemented a new PostgreSQL database for data management.
	- Added configuration files for environment settings and Docker services.

- **Bug Fixes**
	- Adjusted routing for the "Home" menu item in the header component.

- **Documentation**
	- Updated configuration files to outline new environment settings and service dependencies.

### CODE REVIEW

File: .env [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+POSTGRES_PASSWORD=keycloak_db_user_password
+KEYCLOAK_ADMIN=admin
+KEYCLOAK_ADMIN_PASSWORD=password
+KEYCLOAK_CLIENT_ID="nextjs"
+KEYCLOAK_CLIENT_SECRET="gX0nfIp57bO3bY68br3kdTXung2Auwpr"
+KEYCLOAK_ISSUER="http://localhost:8080/realms/shlp"
+NEXTAUTH_URL="http://localhost:3000"
+NEXTAUTH_SECRET="ArmMS4LBmH3VTR77UrSAY2lPs04HO0Nk2/4BcU0jMvc="
+NEXT_CONTAINER_KEYCLOAK_ENDPOINT="http://keycloak:8080"
+NEXT_LOCAL_KEYCLOAK_URL="http://localhost:8080"
```
Comment (BMartinos) [inline .env:12]: Lets move this `.env` data into the `.env.example`. and also copy the env keys into the `.env.production.example`. 

We separate these to have environment specific config defined. 

The `.env` isnt being committed to the repo, we make a copy of the example and populate the details needed per instance

File: src/app/components/Logout.tsx [hunk 2]
```diff
@@ -0,0 +1,8 @@
+"use client"
+import { Button, Typography } from "@mui/material";
+import { signOut } from "next-auth/react";
+export default function Logout() {
+  return <Button onClick={() => signOut()} variant="contained" color="secondary">
+    <Typography>Sign out</Typography>
+  </Button>
+}
```
Comment (BMartinos) [inline src/app/components/Logout.tsx:8]: minor: Lets add some new line separators between the main sections

```suggestion
"use client";

import { Button, Typography } from "@mui/material";
import { signOut } from "next-auth/react";

export default function Logout() {
  return (
    <Button 
      onClick={() => signOut()} 
      variant="contained" 
      color="secondary"
    >
      <Typography>Sign out</Typography>
    </Button>
  );
}
```

File: src/app/components/Login.tsx [hunk 3]
```diff
@@ -0,0 +1,8 @@
+"use client"
+import { Button } from "@mui/material";
+import { signIn } from "next-auth/react";
+export default function Login() {
+  return <Button onClick={() => signIn("keycloak")} variant="contained" color="secondary">
+    Sign in
+  </Button>
+}
```
Comment (BMartinos) [inline src/app/components/Login.tsx:8]: Similar comment around adding in some line separators. This is just to make the code a bit more readable

File: src/app/api/auth/[...nextauth]/route.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        },
+        url: `${process.env.NEXT_LOCAL_KEYCLOAK_URL}/realms/shlp/protocol/openid-connect/auth`,
+      },
+      token: `${process.env.NEXT_CONTAINER_KEYCLOAK_ENDPOINT}/realms/shlp/protocol/openid-connect/token`,
+      userinfo: `${process.env.NEXT_CONTAINER_KEYCLOAK_ENDPOINT}/realms/shlp/protocol/openid-connect/userinfo`,
+    })
+  ]
+}
+const handler = NextAuth(authOptions);
+export { handler as GET, handler as POST }
```
Comment (BMartinos) [inline src/app/api/auth/[...nextauth]/route.ts:23]: Similar comment around adding in some line separators. This is just to make the code a bit more readable

**Review by BMartinos — COMMENTED** [review]
@IkbelTalebHssan @yassinedorbozgithub overall looks good to me. Just some minor comments. 

Only other comment is to remove the package-lock.json, as we are alraedy making use of the yarn.lock. We dont need two locks files

**Comment by IkbelTalebHssan** [comment]
Hi @BMartinos,
All the required changes are applied. May we please ask you to review again ? 
Thank you.

File: .env.example [hunk 5]
```diff
@@ -0,0 +1,12 @@
+POSTGRES_DB=keycloak_db
+POSTGRES_USER=keycloak_db_user
+POSTGRES_PASSWORD=keycloak_db_user_password
+KEYCLOAK_ADMIN=admin
+KEYCLOAK_ADMIN_PASSWORD=password
+KEYCLOAK_CLIENT_ID="nextjs"
+KEYCLOAK_CLIENT_SECRET="gX0nfIp57bO3bY68br3kdTXung2Auwpr"
+KEYCLOAK_ISSUER="http://localhost:8080/realms/shlp"
```
[the remark this answers is not in the document]
Comment (BMartinos) [reply .env.example:8]: This secret is only used for development purposes.
The production secret will be stored in a key manager and injected once needed

**Comment by BMartinos** [comment]
@yassinedorbozgithub @IkbelTalebHssan I think we can merge this PR in now. Just noticed there is a merge conflic in the yarn.lock. If we can just correct this then its good to be merged :+1:

**Comment by yassinedorbozgithub** [comment]
Hi @BMartinos, we resolve myself and @IkbelTalebHssan the yarn.lock file conflicts.
The PR is now ready to be merged :heavy_check_mark:

**Comment by BMartinos** [comment]
Thanks @yassinedorbozgithub Lets proceed to merge this PR into master :+1: :rocket:

**Pull request merged**